### ISSUE TITLE
Datastore: Using new datastore aggregation queries for count

### ISSUE THREAD

**Issue by juniorjainsahab** [GoogleCloudPlatform/spring-cloud-gcp#1781]
In the current implementation, the library performs client-side aggregations by fetching all the keys on the client side and then figuring out the count. In theory, the new aggregation queries and COUNT aggregation should be faster than the current one, as the current implementation makes use of the lazy iterator from the underlying client library that initiates multiple backend calls in cases where the number of keys is greater than the page size specified. And on top of that getting all the keys on the client side will have **more egress cost** compared to getting a count value.

Reference: [COUNT aggregation and Aggregation queries in datastore](https://cloud.google.com/datastore/docs/aggregation-queries)

**Comment by meltsufin** [comment]
@kolea2 Would you mind giving us some feedback on this? This looks like a very reasonable and seamless improvement to the way we do `count()` in the `DatastoreTemplate`. Should we be aware of any caveats or incompatibility with the current `findAllKeys(entityClass).length` aproach?

**Comment by kolea2** [comment]
@meltsufin I'm unfamiliar with `findAllKeys` (seems like custom code in this library), but using `count()` from the Datastore API directly sounds great to me! CC @jainsahab as fyi

**Comment by meltsufin** [comment]
@kolea2 Thanks for feedback. The way we did the count before is this:
https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/a478264935743311ac9bb1d22f95553d317d84a5/spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java#L865-L872

**Comment by jainsahab** [comment]
This is what doc says:
`A keys-only query is counted as a single entity read for the query itself. The individual results are counted as small operations.`
I got confused and was under the impression that for keys only queries users will only be charged **only 1 entity read** regardless the number of keys (millions or billions) returned by that key only query, whereas aggregation query cost will grow linearly `Math.ceil(NUMBER_OF_INDEX_ENTERIES_SCANNED / 1000)`
For ex: count() operations that match between 0 and 1000 index entries are billed for one entity read. For a count() operation that matches 1500 index entries, you are billed 2 entity reads.

But yeah, <ins> **cost should actually be same**</ins>, as the a key only query will get 1000 results per response and will be charged **1 entity read**, and underlying client library will make multiple requests to satisfy the query and that's how pricing will end up being the same.

In a nutshell since aggregation queries do not have any cost implications, they are definitely better over keys only queries `(less egress cost and runtime efficient)`.
